Apps/FlightSoftware/Watchdog2/src/ground_cmd.c

```diff
@@ -0,0 +1,498 @@
+    }
+
+    /* check that the length is greater than the base sane amount */
+    // TODO: pp->used < pp_len ||  check ?
+    if (pp_len < 11) {
+        /* malformed packet length */
+        reply_ground_cmd(0, GNDRESP_EPACKETLEN);
+        return;
+    }
+
+    /* first is sequence number */
+    /* TODO: sequence number */
+    /* next is two bytes, little endian, representing length of the packet */
+    packet_len = buf[2] << 8 | buf[1];
+    /* check that the length is right */
+    if ((packet_len + 4) != pp_len) {
+        /* malformed packet length */
+        reply_ground_cmd(0, GNDRESP_EPACKETLEN);
+        return;
+    }
+
+    /* TODO: check the checksum */
+    (void)(buf[3]);
+
+    /* get the type (should be 000BADA55 for hercules) */
+    if (buf[4] == 0xEE && buf[5] == 0xFF && buf[6] == 0x00 && buf[7] == 0xC0) {
+        /* command */
+        handle_ground_cmd(buf + 8, pp_len - 8);
+    } else {
+        /* forward it on to hercules */
+        // copy the buffer into the hercules buffer
+        memcpy(hercbuf.buf + hercbuf.used, pp->buf, pp->used); // @suppress("Invalid arguments")
+        hercbuf.used += pp->used;
+        if (hercbuf.used > BUFFER_SIZE) {
+            // TODO: better logic here. should overwrite old like a ring buffer, not just
+            // reset from 0
+            hercbuf.used = 0;
+        }
+    }
+
+    /* done parsing */
+    // TODO: multiple packets? handle.
+    pp->used = 0;
+    pp->idx = 0;
+}
+
+/**
+ * Send the earth heartbeat
+ */
+void send_earth_heartbeat() {
+    static uint8_t counter = 0;
+    uint8_t send_buf[32];
+    if (counter % 3 != 2) {
+        // send every 2 seconds
+        counter++;
+        //return;
+    }
+    counter = 0;
+
+    // build the packet
+    send_buf[0] = 0xFF;
+    // TODO: tvac changes
+    // send adc value temperature
+    send_buf[1] = (uint8_t)(adc_values[ADC_TEMP_IDX]);
+    send_buf[2] = (uint8_t)(adc_values[ADC_TEMP_IDX] >> 8);
+
+    // send adc value temperature
+    send_buf[3] = (uint8_t)(raw_battery_charge[0]);
+    send_buf[4] = (uint8_t)(raw_battery_charge[1]);
+
+    // send adc value temperature
+    send_buf[5] = (uint8_t)(raw_battery_voltage[0]);
+    send_buf[6] = (uint8_t)(raw_battery_voltage[1]);
+
+    // send adc value temperature
+    send_buf[7] = (uint8_t)(raw_battery_current[0]);
+    send_buf[8] = (uint8_t)(raw_battery_current[1]);
+
+    // send adc value temperature
+    send_buf[9] = (uint8_t)(raw_fuel_gauge_temp[0]);
+    send_buf[10] = (uint8_t)(raw_fuel_gauge_temp[1]);
+
+    // send ASDF
+    send_buf[11] = (uint8_t)(Kp_heater);
+    send_buf[12] = (uint8_t)(Kp_heater >> 8);
+
+    // send ASDF
+    send_buf[13] = (uint8_t)(heater_setpoint);
+    send_buf[14] = (uint8_t)(heater_setpoint >> 8);
+
+    // send ASDF
+    send_buf[15] = (uint8_t)(heater_window);
+    send_buf[16] = (uint8_t)(heater_window >> 8);
+
+    // send ASDF
+    send_buf[17] = (uint8_t)(PWM_limit);
+    send_buf[18] = (uint8_t)(PWM_limit >> 8);
```

Comment:
I already said this in a Slack message to Mason, but I'll copy it here so all of my comments are in one place.

Serializing/deserializing bytes to/from buffers is something that is done quite frequently across the code base (e.g. here, in `watchdog_send_hercules_telem()` in `watchdog.c`). Repeating it so many times can be error prone. I wrote something way back that may help a little.

I've attached three files to this comment: [Serialization.c.txt](https://github.com/PlanetaryRobotics/CubeRoverPackage/files/6467246/Serialization.c.txt), [Serialization.h.txt](https://github.com/PlanetaryRobotics/CubeRoverPackage/files/6467245/Serialization.h.txt), and [TeleopMessages.c.txt](https://github.com/PlanetaryRobotics/CubeRoverPackage/files/6467247/TeleopMessages.c.txt). All three have `.txt` appended because Github wouldn't let me attach files with the `.c` and `.h` extensions, so if you download these you should remove the `.txt` extension. The Serialization files are the header and source for the code that I think may help here. TeleopMessages.c is just an example of how I used the Serialization interface elsewhere.

I'll give an example of how they could be used here. Since MSP430 is little endian `adc_values[ADC_TEMP_IDX] >> 8`  is equivalent to
```
uint16_t* adc_temp_value = adc_values + ADC_TEMP_IDX;
((uint8_t*) (adc_temp_value)[1]
```
Additionally, it appears `send_buf` is also little endian. This means that the code here could become
```
uint8_t *send_buf_dst = send_buf + 1;

// send adc value temperature
Serialization__serializeAs16Bit(adc_values + ADC_TEMP_IDX, send_buf_dst, SERIALIZATION__LITTLE_ENDIAN);
send_buf_dst += sizeof(adc_values[ADC_TEMP_IDX]);

// send adc value temperature
Serialization__serializeAs16Bit(raw_battery_charge, send_buf_dst, SERIALIZATION__LITTLE_ENDIAN);
send_buf_dst += sizeof(raw_battery_charge);

// send adc value temperature
Serialization__serializeAs16Bit(raw_battery_voltage, send_buf_dst, SERIALIZATION__LITTLE_ENDIAN);
send_buf_dst += sizeof(raw_battery_voltage);

// send adc value temperature
Serialization__serializeAs16Bit(raw_battery_current, send_buf_dst, SERIALIZATION__LITTLE_ENDIAN);
send_buf_dst += sizeof(raw_battery_current);

// send adc value temperature
Serialization__serializeAs16Bit(raw_fuel_gauge_temp, send_buf_dst, SERIALIZATION__LITTLE_ENDIAN);
send_buf_dst += sizeof(raw_fuel_gauge_temp);

// send ASDF
Serialization__serializeAs16Bit(&Kp_heater, send_buf_dst, SERIALIZATION__LITTLE_ENDIAN);
send_buf_dst += sizeof(Kp_heater);

// send ASDF
Serialization__serializeAs16Bit(&heater_setpoint, send_buf_dst, SERIALIZATION__LITTLE_ENDIAN);
send_buf_dst += sizeof(heater_setpoint);

// send ASDF
Serialization__serializeAs16Bit(&heater_window, send_buf_dst, SERIALIZATION__LITTLE_ENDIAN);
send_buf_dst += sizeof(heater_window);

// send ASDF
Serialization__serializeAs16Bit(&PWM_limit, send_buf_dst, SERIALIZATION__LITTLE_ENDIAN);
send_buf_dst += sizeof(PWM_limit);
```
It's not less code, but you're not repeatedly manually shifting and/or indexing (which is inherently error-prone, since you could easily make a copy-paste error or similar). I'm not suggesting you write custom serialization/deserialization functions for every single buffer you read from or write to like I have in my TeleopsMessages file, but you can replace the manual shifting and indexing with Serialization calls.

This isn't a major issue, and nothing is wrong with what you currently have. It's just a change that would make the code slightly less error-prone.